src/engine/effects/engineeffectsdelay.cpp

```diff
@@ -0,0 +1,86 @@
+#include "engine/effects/engineeffectsdelay.h"
+
+#include "util/rampingvalue.h"
+#include "util/sample.h"
+
+namespace {
+// See enginedelay.cpp
+constexpr int kiMaxDelay = static_cast<int>(0.508 *
+        mixxx::audio::SampleRate::kValueMax * mixxx::kEngineChannelCount);
+} // anonymous namespace
+
+EngineEffectsDelay::EngineEffectsDelay()
+        : m_currentDelaySamples(0),
+          m_prevDelaySamples(0),
+          m_delayBufferWritePos(0) {
+    m_pDelayBuffer = SampleUtil::alloc(kiMaxDelay);
+}
+
+void EngineEffectsDelay::process(CSAMPLE* pInOut,
+        const int iBufferSize) {
+    if (m_prevDelaySamples == 0 && m_currentDelaySamples == 0) {
+        return;
+    }
+
+    int delaySourcePos =
+            (m_delayBufferWritePos + kiMaxDelay - m_currentDelaySamples) %
+            kiMaxDelay;
+
+    VERIFY_OR_DEBUG_ASSERT(delaySourcePos >= 0) {
+        return;
+    }
+    VERIFY_OR_DEBUG_ASSERT(delaySourcePos <= kiMaxDelay) {
+        return;
+    }
+
+    if (m_prevDelaySamples == m_currentDelaySamples) {
+        for (int i = 0; i < iBufferSize; ++i) {
+            // Put samples into delay buffer.
+            m_pDelayBuffer[m_delayBufferWritePos] = pInOut[i];
+            m_delayBufferWritePos = (m_delayBufferWritePos + 1) % kiMaxDelay;
```

Comment:
Do you plan to get rid of this % ? I would prefer a memcpy() based solution:
https://github.com/mixxxdj/mixxx/blob/b58dd5bc990a1f6afe7a400e9b9451882c78112a/src/util/sample.h#L53

Reply:
IMO this requires some more complicated code to be safe, but it would be a worthwhile optimization too look into.

Reply:
This may serve as a model: 
https://github.com/mixxxdj/mixxx/blob/b58dd5bc990a1f6afe7a400e9b9451882c78112a/lib/portaudio/pa_ringbuffer.c#L198
copy in two steps form the current position to the ring wrap around and then the remaining samples.

Reply:
Thank you very much for this tip. It seems like it could be really helpful.